### PR TITLE
bcm2711: Add USB_LAN78XX Pi Compute Module 4 Carrier Boards

### PR DESCRIPTION
**Pull request by geerlingguy**
I'm trying to get OpenWRT installed on a [Seeed Studios Raspberry Pi Compute Module 4 Carrier Board](https://www.seeedstudio.com/Rapberry-Pi-CM4-Dual-GbE-Carrier-Board-p-4874.html) with two Gigabit network interfaces.

One interface uses the built-in Gigabit NIC on the Compute Module 4, and it works fine with the current BCM2711 defaults. The other, however, uses the Microchip LAN7800 USB Gigabit controller that's also found on the Pi 3 B+, and I noticed that while the BCM2709 config includes the LAN78XX module, the BCM2711 does not, so this board would require a custom kernel to work with both interfaces and OpenWRT.

This patch adds in LAN78XX support so the Seeed Studios router board should work with OpenWRT out of the box.

### CODE REVIEW

**Comment by bobafetthotmail** [comment]
Most of what you wrote in the PR message on github should be put in the commit message instead (and lines cut to 75 char lenght), and also what you used in the PR title should also go in the commit subject. 
There is no Signed-off-by in the commit.
Please see submission rules for more details https://openwrt.org/submitting-patches

**Comment by geerlingguy** [comment]
Will reformat in a bit.

**Comment by geerlingguy** [comment]
@bobafetthotmail - Reformatted the commit. Thanks!

**Comment by adschm** [comment]
Missing prefix for commit message. Apart from it, the commit message is wrong, since you say you add a specific config symbol but you actually add two ...

**Comment by geerlingguy** [comment]
@adschm - Added the module to the commit title, and a mention of `CONFIG_USB_NET_DRIVERS` to the commit message (didn't know if that was obvious or not w/r/t the dependency).

**Comment by wulfy23** [comment]
hi jeff...

can you confirm ethernet ordering for these module(s?) with patch applied...?

i.e. is lan consistently @ eth0 ?

**Comment by geerlingguy** [comment]
@wulfy23 - Yes, in the two times I've tested, onboard LAN was still `eth0`.

**Comment by wulfy23** [comment]
per https://github.com/openwrt/openwrt/pull/4216#issuecomment-855310571

consider alternative implementation similar to below;
```
diff --git a/target/linux/bcm27xx/image/Makefile b/target/linux/bcm27xx/image/Makefile
index f0c3ae1272..3877134660 100644
--- a/target/linux/bcm27xx/image/Makefile
+++ b/target/linux/bcm27xx/image/Makefile
@@ -158,7 +158,18 @@ define Device/rpi-4
   DEVICE_PACKAGES := \
        cypress-firmware-43455-sdio \
        cypress-nvram-43455-sdio-rpi-4b \
-       kmod-brcmfmac wpad-basic-wolfssl
+      kmod-brcmfmac \
+      kmod-usb-core kmod-mii kmod-usb-net \
+      kmod-libphy \
+      kmod-usb-net-aqc111 \
+      kmod-usb-net-asix \
+      kmod-usb-net-asix-ax88179 \
+      kmod-usb-net-pegasus \
+      kmod-usb-net-rtl8152 \
+      kmod-usb-net-rtl8150 \
+      kmod-usb-net-sr9700 \
+      kmod-usb-net-smsc95xx \
+      wpad-basic-wolfssl
   IMAGE/sysupgrade.img.gz := boot-common | boot-2711 | sdcard-img | gzip | append-metadata
   IMAGE/factory.img.gz := boot-common | boot-2711 | sdcard-img | gzip
 endef
```

**Comment by rhester72** [comment]
This would be *hugely* helpful for out-of-box bring-up.

**Comment by Noltari** [comment]
@geerlingguy  I don't want to enable additional symbols which aren't needed for the _basic_ RPi models.
I would prefer to add additional specific CM4 profiles which include the needed kernel packages for those boards (such as kmod-usb-net-lan78xx, which would need to be created as @wulfy23 pointed out).

**Comment by sergey-brutsky** [comment]
Hey @geerlingguy do you have any plans to finish this PR ?

Until that work is done we need to use dirty tricks like [this one](https://github.com/sergey-brutsky/openwrt-seeed-carrier-board) to to unlock the potential of seeed carrier boards.

**Comment by geerlingguy** [comment]
@sergey-brutsky - Sorry, unfortunately right now I don't have the time; I will be digging back into some OpenWRT work in the next couple months, but can't guarantee I'd be able to work on the PR further before then. I would suggest either branching off this PR or creating a new one if someone else wants to bring the feature all the way home :)

**Comment by BKPepe** [comment]
@Noltari:

> I would prefer to add additional specific CM4 profiles 

I think this approach would be more complicated for users. Doing that just for two packages does not make any sense from my point of view. It won't be very clear in the end.

> which include the needed kernel packages for those boards (such as kmod-usb-net-lan78xx, which would need to be created as @wulfy23 pointed out).

Well... @geerlingguy did it exactly the same as it is done for BCM2709 and it is mentioned that in the commit message as well. 
Reference:  https://github.com/openwrt/openwrt/blob/32c74552b2310531af593ee0f859a3935a3a7cbd/target/linux/bcm27xx/bcm2709/config-5.4#L437

For me, it seems like you are now forcing a new external contributor to do the changes, which should be done in the first place. Don't you think? He does not have any knowledge or experience in building OpenWrt packages.

I received [Seeed Studios Raspberry Pi Compute Module 4 Carrier Board](https://www.seeedstudio.com/Rapberry-Pi-CM4-Dual-GbE-Carrier-Board-p-4874.html) , too a few days ago. I prepared a build system and created the package ``kmod-usb-net-lan78xx`` and in the meantime, I wondered why no one added it to the official OpenWrt image, so I started looking at found these two issues, which led me to this PR.

- https://github.com/geerlingguy/raspberry-pi-pcie-devices/issues/137
- https://github.com/Seeed-Studio/seeed-linux-openwrt/issues/21

I am not surprised that almost anyone is trying to fork the repository and don't contribute changes back because this PR has been sitting here for how long? Almost a year. Too bad. :(

TL;DR: Things should not be complicated like this, and it should be more user-friendly. Two kernel options for RPI4 CM4 do not harm anyone to have it enabled by default. How much storage does RPI4 CM have? Well, you can choose if you want to have microSD card or eMMC.

**Comment by wulfy23** [comment]
" Things should not be complicated like this, and it should be more user-friendly. Two kernel options for RPI4 CM4 do not harm anyone to have it enabled by default.

If a new 'Fancy Audio DAC" carrier board is created, shall we add this symbol to all CM4/RPI4/400 images too?

" wondered why no one added it to the official OpenWrt image

because;
- the vendor did not bother to create a 10 line PR adding the ipk
- no hardware on these boards used this chip when they were added
- technically the hardware is not on the board itself, we can have 50 new ones pop up tommorrow, each with a total of 3-5 users each

" @geerlingguy did it exactly the same as it is done for BCM2709

this is a chip on the OEM board vs a carrier... it's not the same thing all users of 2709 NEED it, all users of rpi4/400/cm4 don't

have you asked seeed why they have not bothered to send in a PR for the ipk for their carrier board?

**Comment by BKPepe** [comment]
Oh @wulfy23, I would say that you misunderstand it. Let me rephrase it for you.

I have [RPI CM4 from Seeed studio](https://www.seeedstudio.com/Dual-GbE-Carrier-Board-with-4GB-RAM-32GB-eMMC-RPi-CM4-Case-p-5029.html), it has 2 Ethernet ports. One of the ports uses: ``Microchip's LAN7800 USB 3.0 to GbE``. Don't you think that you expect both of the ports to work out the box for a router? It is crucial to have WAN and LAN ports working, right? Right now, there is working only LAN. 

> If a new 'Fancy Audio DAC' carrier board is created, shall we add this symbol to all CM4/RPI4/400 images too?

My paragraph above explains it. There are important things, which you need to have working on your device. If you don't, you can not download additional packages, you can use the ``USB flash drive`` to install packages or ``SCP`` to get the device working, so it should be able to connect to the Internet. Without working Internet, you can not do anything.

Is Fancy Audio DAC so important that the driver should be installed by default? Of course, it should not. If you do have working Ethernet ports, you can download it. On the other hand, it would be nice to have something like Turris has. It is called [Updater drivers](https://gitlab.nic.cz/turris/os/updater-lists/-/tree/master/drivers), so if you have working Internet (which is imho important) and you will insert/plug device from the recognized list, it will automatically download the drivers. It improves the user experience, but we are getting off-topic on this one.

Overall, you didn't give a good example.

> no hardware on these boards used this chip when they were added

Ehm... The hardware (if we are talking about lan78xx) is used on Raspberry Pi 3B+.

___

Based on your arguing, we can ask OpenWrt developers why this PR and other PRs are stuck, right? Just take a look at umbim/wwan france here. 4 pull requests, still not merged.

**Review by BKPepe — APPROVED** [review]

**Comment by wulfy23** [comment]
important thing is you pushed the PR to add the kmod, thats the main thing that held up the PR... this gets most people with the device up and running with imagebuilder and asu...

don't think OP is keen on tackling the intricacies of image/Makefile variants... so we should probably take this to a new PR that does that or to the mailing list if more discussion is to be had about the default package set for these ARM largeflash devices...

and/or seek some creative ways to handle variant definitions or similar alternatives...

thanks for getting the ball rolling again...

**Review by sergey-brutsky — APPROVED** [review]
LGTM

**Comment by Noltari** [comment]
This is no longer needed thanks to https://github.com/openwrt/openwrt/pull/9494 (31ccc27f41f6844f91728c7448d6ca0a0d0ae5b5).
